source/SH3/graphics/texture.cpp

```diff
@@ -13,48 +13,330 @@
+            {
+                pixel = palette[iBuffer[pindex]];
+
+                data[doff]      = pixel.r;
+                data[doff + 1]  = pixel.g;
+                data[doff + 2]  = pixel.b;
+            }
+        }
+        else // If the distortion flag isn't set, just read the pixel data in from the palette.
+        {
+            for(int i = 0; i < header.texWidth * header.texHeight * 3; i += 3)
+            {
+                file.ReadData(&index, sizeof(index), e);
+                pixel = palette[index];
+
+                data[i]       = pixel.r;
+                data[i + 1]   = pixel.g;
+                data[i + 2]   = pixel.b;
+            }
+        }
+
+        Dump2Bitmap(header.texWidth, header.texHeight, data);
+    }
+    else if(header.bpp == PixelFormat::RGBA)
+    {
+        file.ReadData(&data[0], header.texSize / sizeof(rgba), e);
+    }
+    else if(header.bpp == PixelFormat::BGR)
+    {
+        file.ReadData(&data[0], header.texSize / sizeof(bgr), e);
+    }
+    else if(header.bpp == PixelFormat::RGBA16)
+    {
+        //TODO: Some kind of fucked up shit here. I think this is R5G5B5A1 or something like that..
+    }
+    else
+    {
+        die("sh3_texture::Load( ): Unknown Pixel Format, %d", header.bpp);
     }
 
-    ret = file.ReadData(&data[0], header.texSize, e);
-    if(ret != header.texSize)
-        Log(LogLevel::WARN, "sh3_texture::Load( ): Only read %d out of %d bytes!", ret, header.texSize);
 
-    // Now actually create a logical texture with OpenGL on the gpu
-    glGenTextures(1, &tex);
+    glGenTextures(1, &tex);             // Create a texture
+    glBindTexture(GL_TEXTURE_2D, tex);  // Bind it for use
 
-    glBindTexture(GL_TEXTURE_2D, tex);
-    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, header.texWidth, header.texHeight, 0, GL_BGRA, GL_UNSIGNED_BYTE, &data[0]);
-    glBindTexture(GL_TEXTURE_2D, 0); // Unbind this texture
+    // Create the texture according to its pixel format!
+    switch(header.bpp)
+    {
+        case PixelFormat::RGBA:     // Regular 32-bit RGBA
+            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, header.texWidth, header.texHeight, 0, GL_RGBA, GL_UNSIGNED_BYTE, &data[0]);
+            break;
+        case PixelFormat::BGR:      // 24-bit BGR
+            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, header.texWidth, header.texHeight, 0, GL_BGR, GL_UNSIGNED_BYTE, &data[0]);
+            break;
+        case PixelFormat::RGBA16:   // 16-bit RGBA. OpenGL supports this (I think)
+            // We need to split this up ourselves. Not quite sure how it's stored...
+            break;
+        case PixelFormat::PALETTE:
+            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, header.texWidth, header.texHeight, 0, GL_RGB, GL_UNSIGNED_BYTE, &data[0]);
+            break;
+        default:
+            die("sh3_texture::Load( ): Invalid pixel format: %d", header.bpp);
+            break;
+    }
+
+//    glBindTexture(GL_TEXTURE_2D, tex); // Un-bind this texture.
```

Comment:
I commented the whole function before I rewrote it. This missed my selection before uncomment haha.